client-core/src/wallet/syncer.rs

```diff
@@ -0,0 +1,441 @@
+#![allow(missing_docs)]
+use itertools::{izip, Itertools};
+use secstr::SecUtf8;
+use std::sync::mpsc::Sender;
+
+use chain_core::common::H256;
+use chain_core::tx::data::TxId;
+use client_common::tendermint::types::{Block, Status};
+use client_common::tendermint::Client;
+use client_common::{
+    BlockHeader, Error, ErrorKind, PublicKey, Result, ResultExt, SecureStorage, Transaction,
+};
+
+use super::syncer_logic::{handle_block, prepare_block_header};
+use crate::service;
+use crate::service::{KeyService, SyncState, Wallet, WalletState};
+use crate::TransactionObfuscation;
+
+/// A struct for providing progress report for synchronization
+#[derive(Debug)]
+pub enum ProgressReport {
+    /// Initial report to send start/finish heights
+    Init {
+        /// Name of wallet
+        wallet_name: String,
+        /// Block height from which synchronization started
+        start_block_height: u64,
+        /// Block height at which synchronization will finish
+        finish_block_height: u64,
+    },
+    /// Report to update progress status
+    Update {
+        /// Name of wallet
+        wallet_name: String,
+        /// Current synchronized block height
+        current_block_height: u64,
+    },
+}
+
+/// Transaction decryptor interface for wallet synchronizer
+pub trait TxDecryptor: Clone + Send + Sync {
+    /// decrypt transaction
+    fn decrypt_tx(&self, txids: &[TxId], pubkey: &PublicKey) -> Result<Vec<Transaction>>;
+}
+
+impl<F> TxDecryptor for F
+where
+    F: Fn(&[TxId], &PublicKey) -> Result<Vec<Transaction>> + Clone + Send + Sync,
+{
+    fn decrypt_tx(&self, txids: &[TxId], viewkey: &PublicKey) -> Result<Vec<Transaction>> {
+        self(txids, viewkey)
+    }
+}
+
+/// Load view key and decrypt transaction with `TransactionObfuscation` and `KeyService`
+#[derive(Clone)]
+pub struct TxObfuscationDecryptor<O: TransactionObfuscation, S: SecureStorage> {
```

Comment:
I guess it's all right to separate this out, containing all "external state things" needed as inputs for  `TransactionObfuscation`.
I'd need to look at it in more detail what its life-span is and whether it makes sense to hold all those things

Reply:
> I guess it's all right to separate this out, containing all "external state things" needed as inputs for `TransactionObfuscation`.
> I'd need to look at it in more detail what its life-span is and whether it makes sense to hold all those things

`TxDecryptor` is mainly useful when doing the mocking test, you can simply write a closure to provide the functionality.
Another thing is, in the future, the passphrase stored in `TxObfuscationDecryptor` might be different from the passphrase used to store `WalletState`.

Reply:
Actually, since there is only one viewkey per wallet, we should remove the `viewkey` paramter from `decrypt_tx`. And `TxObfuscationDecryptor` can hold the private view key in memory, rather than load and decrypt every time.